File: src/main/java/com/app/umkaSchool/model/ExerciseType.java
package com.app.umkaSchool.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.time.ZonedDateTime;
import java.util.UUID;

@Data
@Entity
@Table(name = "exercise_type", schema = "school")
public class ExerciseType {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "exercise_type_id", nullable = false)
    private UUID id;

    @Column(nullable = false)
    private String name;

    private String description;

    @Column(name = "base_difficulty", nullable = false)
    private Integer baseDifficulty;

    @Column(name = "avg_time_seconds", nullable = false)
    private Integer avgTimeSeconds;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "parameter_ranges", columnDefinition = "jsonb")
    private String parameterRanges; // JSON: {"cardCount": [2, 20], "displaySpeed": [0.5, 3.0], "timePerQuestion": [2, 20]}

    @JsonIgnore
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private Teacher createdBy;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private ZonedDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private ZonedDateTime updatedAt;
}

File: frontend/src/types/exerciseType.ts
export interface ParameterRanges {
    cardCount?: [number, number];
    displaySpeed?: [number, number];
    timePerQuestion?: [number, number];
    exampleCount?: [number, number];
    dividendDigits?: [number, number];
    divisorDigits?: [number, number];
    firstMultiplierDigits?: [number, number];
    minValue?: number;
    maxValue?: number;
    digitTypes?: string[];
    themes?: string[];
}

export interface ExerciseType {
    id: string;
    name: string;
    description: string;
    difficulty: 'beginner' | 'intermediate' | 'advanced';
    duration: string;
    parameterRanges?: ParameterRanges;
}

export interface APIExerciseType {
    id: string;
    name: string;
    description: string;
    baseDifficulty: number;
    avgTimeSeconds: number;
    parameterRanges?: string; // JSON string
    createdById: string;
    createdByName: string;
    createdAt: string;
    updatedAt: string;
}

